src/SAWScript/MGU.hs

```diff
@@ -1300,32 +1324,78 @@ inferSingleStmt ln pos ctx s = do
 
 -- Create a type schema for a declaration out of its free vars.
 --
+-- (actually for a list of declarations)
+--
 -- (This creates names for any remaining unification vars, so
 -- potentially updates the expression.)
 generalize :: [OutExpr] -> [Type] -> TI [(OutExpr,Schema)]
-generalize es0 ts0 =
-  do es <- applyCurrentSubst es0
+generalize es0 ts0 = do
+     -- first, substitute away any resolved unification variables
+     -- in both the expressions and types.
+     es <- applyCurrentSubst es0
      ts <- applyCurrentSubst ts0
 
-     envUnify <- unifyVarsInEnv
-     envNamed <- namedVarsInEnv
-     let is = M.toList (unifyVars ts M.\\ envUnify)
-     let bs = M.toList (namedVars ts M.\\ envNamed)
+     -- Extract lists of any unification vars and named type vars that
+     -- still appear.
+     let is = unifyVars ts
+     let bs = namedVars ts
+
+     -- Drop any unification vars and named type vars that we
+     -- shouldn't forall-bind.
+     --
+     -- For unification vars, any whose scope reaches beyond the
+     -- current declaration should be left alone; they should only be
+     -- bound when they eventually move out of scope. Get these by
+     -- examining the types used in the right-hand sides of both the
+     -- variable environment and the type environment.
+     --
+     -- For named vars, exclude any that appear that appear as keys
+     -- (on the left-hand side) of the type environment. Those are
+     -- already defined.
+     --
+     -- There should be no other named variables involved. We don't
+     -- allow referring to random unbound type names any more, and
+     -- we don't yet (though probably should, FUTURE) have a way of
+     -- explicitly forall-binding type variables in declarations.
+     --
+     -- It would be handy for scaling if we didn't have to examine
+     -- the entire variable environment (on the grounds that there
+     -- should be no loose unification vars at the top level where
+     -- most definitions will come from) but (a) we don't have the
+     -- structure to support that and (b) it is not absolutely clear
+     -- that there isn't a way to get such loose unification vars,
+     -- in which case we'd have to do something about it.
+     --
+     -- This code also used to exclude named vars used on the
+     -- right-hand side of the variable environment; this was to allow
+     -- the use of otherwise undefined type names in the primitives
+     -- table. There is no longer any need for such hackery, and
+     -- undefined type names are not allowed to appear in the variable
+     -- environment.
+
+     envUnifyVars <- unifyVarsInEnvs
+     knownNamedVars <- namedVarDefinitions
+     let is' = is M.\\ envUnifyVars
+     let bs' = bs M.\\ knownNamedVars
+
+     -- convert to lists
+     let is'' = M.toList is'
+     let bs'' = M.toList bs'
```

Comment:
A minor thing, but I would prefer using numbers as suffixes rather than attaching arbitrary numbers of apostrophes. That is, I would prefer `is1`, `is2`, `is3` over `is'`, `is''`, `is'''`, etc., as the more apostrophes you attach, the more you have to squint to figure out which one you're referring to.